packages/vite/src/node/build.ts

```diff
@@ -674,14 +677,24 @@ export function resolveLibFilename(
     return libOptions.fileName(format)
   }
 
-  const name = libOptions.fileName || getPkgName(root)
+  const packageJson = getPkgJson(root)
+  const name = libOptions.fileName || getPkgName(packageJson.name)
 
   if (!name)
     throw new Error(
       'Name in package.json is required if option "build.lib.fileName" is not provided.'
     )
 
-  return `${name}.${format}.js`
+  let extension: string
+  const isEsFormat = format === 'es'
+
+  if (packageJson?.type === 'module') {
+    extension = isEsFormat ? 'js' : 'cjs'
+  } else {
+    extension = isEsFormat ? 'mjs' : 'js'
+  }
```

Comment:
```suggestion
  if (packageJson?.type === 'module') {
    extension = format === 'cjs' ? 'cjs' : 'js'
  } else {
    extension = format === 'es' ? 'mjs' : 'js'
  }
```

I think something like this should be more accurate, for cases where we have the `iife` or `umd` formats, they should end with `.js`.

Reply:
This makes sense for `iife` but if you externalize a package `umd` will have the `require` call for that package.